pkg/operator/controller/dnsnameresolver/controller.go

```diff
@@ -0,0 +1,217 @@
+	if err := c.Watch(source.Kind(corednsEndpointsSliceCache, &discoveryv1.EndpointSlice{}), handler.EnqueueRequestsFromMapFunc(skipReconcileForEndpointSlice)); err != nil {
+		return nil, nil, err
+	}
+
+	return c, []cache.Cache{dnsNameResolverCache, corednsEndpointsSliceCache}, nil
+}
+
+// skipReconcileForEndpointSlice just skips the reconcile if any event related to the CoreDNS pod
+// endpointslices is received.
+func skipReconcileForEndpointSlice(context.Context, client.Object) []reconcile.Request {
+	return nil
+}
+
+type Config struct {
+	OperandNamespace         string
+	ServiceName              string
+	DNSNameResolverNamespace string
+}
+
+// reconciler handles the actual DNSNameResolver reconciliation logic in response to events.
+type reconciler struct {
+	config Config
+
+	dnsNameResolverCache cache.Cache
+	client               client.Client
+	res                  *Resolver
+	startResolver        sync.Once
+}
+
+// Reconcile expects request to refer to an DNSNameResolver resource, and will do all the work to
+// keep the status of the resource updated.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	// Ensure that the resolver is started before doing anything else.
+	r.startResolver.Do(func() {
+		r.res.Start()
+	})
+
+	logrus.Infof("reconciling request: %v", request)
+
+	// Get the DNSNameResolver resource.
+	dnsNameResolver := &networkv1alpha1.DNSNameResolver{}
+	if err := r.dnsNameResolverCache.Get(ctx, request.NamespacedName, dnsNameResolver); err != nil {
+
+		// Check if the DNSNameResolver resource is deleted. If so, delete DNS names matching the DNSNameResolver resource
+		// from the resolver.
+		if errors.IsNotFound(err) {
+			r.res.Delete(request.Name)
+			return reconcile.Result{}, nil
+		}
+
+		return reconcile.Result{}, err
+	}
+
+	// Get a copy of the DNSNameResolver resource.
+	dnsNameResolver = dnsNameResolver.DeepCopy()
+
+	// Check if the grace period is over for some of the IP addresses after the expiration of their respective TTLs. If so,
+	// remove those IP addresses and update the status of the resource.
+	if removalOfIPsRequired(&dnsNameResolver.Status) {
+		if err := r.client.Status().Update(ctx, dnsNameResolver); err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// Check if the TTLs of some of the IP addresses have expired. If so, requeue the reconcile request after the minimum remaining
+	// time until the grace period gets over among that of the IP addresses with expired TTLs.
+	if ttlExpired, remainingTime := reconcileRequired(&dnsNameResolver.Status); ttlExpired {
+		return reconcile.Result{Requeue: true, RequeueAfter: remainingTime}, nil
+	}
+
+	dnsName := string(dnsNameResolver.Spec.Name)
+	matchesRegular := !isWildcard(dnsName)
+
+	// Add the DNS name in the spec of the DNSNameResolver resource to the resolver with empty resolved addresses. This is done
```

Comment:
What is the "resolver with empty resolved addresses"?  Please explain more why this needs to be done.

Reply:
> What is the "resolver with empty resolved addresses"? 

When the DNSNameResolver object is created with a DNS name in the spec, the corresponding IP addresses of the DNS name are not known. The resolver needs to send the DNS resolution request for the DNS name so that the status of the object gets updated with the corresponding IP addresses. For this reason the Add function is called with the DNS name and empty resolved addresses slice.

 

> Please explain more why this needs to be done in the reconcile step.

During the reconcile step we will fetch the DNSNameResolver object for which the reconcile was triggered and we can then start the DNS resolution process if it was a create event.

Reply:
Moved to openshift/coredns-ocp-dnsnameresolver#2